### PR TITLE
Add an option to the CMake build to disable building shared libraries

### PR DESCRIPTION
**Pull request by simpkins**
Summary:
Add an option to explicitly disable building shared versions of the
RocksDB libraries.  The shared libraries cannot be built in cases where
some dependencies are only available as static libraries.  This allows
still building RocksDB in these situations.

### CODE REVIEW

**Comment by simpkins** [comment]
The build failure reported is just the cmake-mingw build, which seems flaky on master.

**Comment by siying** [comment]
CC @tchaikov any thoughts on it?

**Comment by siying** [comment]
Also, in case @burtonli sees any problem.

**Comment by simpkins** [comment]
Note that this change by itself does not change the default build behavior.  This just adds a new flag to the CMake build to disable building shared libraries.  Unless you explicitly run CMake with `-DROCKSDB_BUILD_SHARED=OFF` the build behavior stays the same.

However in the future I do think it would would be nice to change the default build behavior to only build static libraries by default  It sounds like RocksDB does not really provide ABI compatibility guarantees between releases, even for minor version bumps.  Given that there are no ABI compatibility guarantees for the shared libraries it seems safer to only build and install static libraries by default.  Building static-only by default would then also make it a little easier for RocksDB to take an optional dependency on Facebook's folly library, which also builds static-only by default.

**Comment by ajkr** [comment]
The policy as I understood it was to provide ABI compatibility across patch versions (i.e., the "z" in "x.y.z"). That way shared library users can get bug fixes without recompiling their applications. I have no idea how valuable this is but just wanted to point out there was some progress in ABI compatibility guarantees.

File: CMakeLists.txt [hunk 1]
```diff
@@ -778,40 +779,44 @@ endif()
...
+    set_target_properties(${ROCKSDB_SHARED_LIB} PROPERTIES
+                          LINKER_LANGUAGE CXX
+                          VERSION ${rocksdb_VERSION}
+                          SOVERSION ${rocksdb_VERSION_MAJOR}
+                          CXX_STANDARD 11
+                          OUTPUT_NAME "rocksdb")
   endif()
 endif()
 
+if(ROCKSDB_BUILD_SHARED AND NOT WIN32)
```
Comment (ajkr) [inline CMakeLists.txt]: Unrelated to this PR, but I do not understand why static is used on WIN32 in all cases except the c tests.

**Comment by siying** [comment]
I'm ready to land it but now there is some conflicts. @simpkins can you help resolve the conflict? I can merge after that.

File: CMakeLists.txt [hunk 2]
```diff
@@ -778,38 +777,36 @@ endif()
 
 if(WIN32)
   set(SYSTEM_LIBS ${SYSTEM_LIBS} shlwapi.lib rpcrt4.lib)
-  set(LIBS ${ROCKSDB_STATIC_LIB} ${THIRDPARTY_LIBS} ${SYSTEM_LIBS})
+  set(ROCKSDB_LIB ${ROCKSDB_STATIC_LIB})
```
Comment (tchaikov) [inline CMakeLists.txt]: this change is included by https://github.com/facebook/rocksdb/pull/5916/commits/11273253f0bd62f4eeb485acf6709cb65f89dd62 already

File: tools/CMakeLists.txt [hunk 3]
```diff
@@ -15,7 +15,7 @@ foreach(src ${TOOLS})
 endforeach()
 
 add_executable(db_stress${ARTIFACT_SUFFIX} db_stress.cc db_stress_tool.cc)
-target_link_libraries(db_stress${ARTIFACT_SUFFIX} ${LIBS})
+target_link_libraries(db_stress${ARTIFACT_SUFFIX} ${ROCKSDB_LIB})
```
Comment (tchaikov) [inline tools/CMakeLists.txt]: could you squash this change into "Minor clean-up to library definitions in CMakeLists.txt"?

Comment (simpkins) [reply tools/CMakeLists.txt]: My understanding from past PRs is that the RocksDB team prefers if pull requests aren't advanced by non-fast-forward commits.  The entire change will be squashed into a single commit when it lands anyway.

**Pull request closed without merging**